SU2_CFD/src/drivers/CDriver.cpp

```diff
@@ -1137,170 +1101,56 @@ void CDriver::Inlet_Preprocessing(CSolver ***solver, CGeometry **geometry,
-    case FEM_LES : fem_ns = true; break;
-    case HEAT_EQUATION: heat = true; break;
-    case FEM_ELASTICITY: fem = true; break;
-    case ADJ_EULER : euler = true; adj_euler = true; break;
-    case ADJ_NAVIER_STOKES : ns = true; turbulent = (config->GetKind_Turb_Model() != NONE); adj_ns = true; break;
-    case ADJ_RANS : ns = true; turbulent = true; adj_ns = true; adj_turb = (!config->GetFrozen_Visc_Cont()); break;
-    case DISC_ADJ_EULER: case DISC_ADJ_INC_EULER: euler = true; disc_adj = true; break;
-    case DISC_ADJ_NAVIER_STOKES: case DISC_ADJ_INC_NAVIER_STOKES: ns = true; disc_adj = true; heat = config->GetWeakly_Coupled_Heat(); break;
-    case DISC_ADJ_RANS: case DISC_ADJ_INC_RANS: ns = true; turbulent = true; disc_adj = true; disc_adj_turb = (!config->GetFrozen_Visc_Disc()); heat = config->GetWeakly_Coupled_Heat(); break;
-    case DISC_ADJ_FEM_EULER: fem_euler = true; disc_adj = true; break;
-    case DISC_ADJ_FEM_NS: fem_ns = true; disc_adj = true; break;
-    case DISC_ADJ_FEM_RANS: fem_ns = true; turbulent = true; disc_adj = true; disc_adj_turb = (!config->GetFrozen_Visc_Disc()); break;
-    case DISC_ADJ_FEM: fem = true; disc_adj_fem = true; break;
-    case DISC_ADJ_HEAT: heat = true; disc_adj_heat = true; break;
-
-  }
-
-  /*--- Determine the kind of FEM solver used for the flow. ---*/
-
-  switch( config->GetKind_FEM_Flow() ) {
-    case DG: fem_dg_flow = true; break;
-  }
-
-  /*--- Load restarts for any of the active solver containers. Note that
-   these restart routines fill the fine grid and interpolate to all MG levels. ---*/
+  /*--- Restart direct solvers. ---*/
 
   if (restart || restart_flow) {
-    if (euler || ns) {
-      SU2_OMP_PARALLEL_(if(solver[MESH_0][FLOW_SOL]->GetHasHybridParallel()))
-      solver[MESH_0][FLOW_SOL]->LoadRestart(geometry, solver, config, val_iter, update_geo);
-      END_SU2_OMP_PARALLEL
-    }
-    if (NEMO_euler || NEMO_ns) {
-      solver[MESH_0][FLOW_SOL]->LoadRestart(geometry, solver, config, val_iter, update_geo);
-    }
-    if (turbulent) {
-      SU2_OMP_PARALLEL_(if(solver[MESH_0][TURB_SOL]->GetHasHybridParallel()))
-      solver[MESH_0][TURB_SOL]->LoadRestart(geometry, solver, config, val_iter, update_geo);
-      END_SU2_OMP_PARALLEL
-    }
-    if (config->AddRadiation()) {
-      solver[MESH_0][RAD_SOL]->LoadRestart(geometry, solver, config, val_iter, update_geo);
-    }
-    if (fem) {
-      if (time_domain) {
-        if (config->GetRestart()) val_iter = config->GetRestart_Iter()-1;
-        else val_iter = config->GetUnst_AdjointIter()-1;
+    for (auto iSol = 0u; iSol < MAX_SOLS; ++iSol) {
+      auto sol = solver[MESH_0][iSol];
+      if (sol && !sol->GetAdjoint()) {
+        /*--- Note that the mesh solver always loads the most recent file (and not -2). ---*/
+        SU2_OMP_PARALLEL_(if(sol->GetHasHybridParallel()))
+        sol->LoadRestart(geometry, solver, config, val_iter + (iSol==MESH_SOL && dt_step_2nd), update_geo);
+        END_SU2_OMP_PARALLEL
```

Comment:
Awesome cleanup of the restart logic🧹

Reply:
Perhaps the only advantage of the solver container being an array.